mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraQuotaCurrentValueDao.java

```diff
@@ -0,0 +1,192 @@
+import com.datastax.oss.driver.api.querybuilder.delete.Delete;
+import com.datastax.oss.driver.api.querybuilder.select.Select;
+import com.datastax.oss.driver.api.querybuilder.update.Update;
+import com.google.common.base.MoreObjects;
+
+import reactor.core.publisher.Mono;
+
+public class CassandraQuotaCurrentValueDao {
+
+    public static class QuotaKey {
+
+        public static QuotaKey of(QuotaComponent component, Username identifier, QuotaType quotaType) {
+            return new QuotaKey(component, identifier, quotaType);
+        }
+
+        private final QuotaComponent quotaComponent;
+        private final Username identifier;
+        private final QuotaType quotaType;
+
+        public QuotaComponent getQuotaComponent() {
+            return quotaComponent;
+        }
+
+        public Username getIdentifier() {
+            return identifier;
+        }
+
+        public QuotaType getQuotaType() {
+            return quotaType;
+        }
+
+        private QuotaKey(QuotaComponent quotaComponent, Username identifier, QuotaType quotaType) {
+            this.quotaComponent = quotaComponent;
+            this.identifier = identifier;
+            this.quotaType = quotaType;
+        }
+
+        @Override
+        public final int hashCode() {
+            return Objects.hash(quotaComponent, identifier, quotaType);
+        }
+
+        @Override
+        public final boolean equals(Object o) {
+            if (o instanceof QuotaKey) {
+                QuotaKey other = (QuotaKey) o;
+                return Objects.equals(quotaComponent, other.quotaComponent)
+                    && Objects.equals(identifier, other.identifier)
+                    && Objects.equals(quotaType, other.quotaType);
+            }
+            return false;
+        }
+
+        public String toString() {
+            return MoreObjects.toStringHelper(this)
+                .add("quotaComponent", quotaComponent)
+                .add("identifier", identifier)
+                .add("quotaType", quotaType)
+                .toString();
+        }
+    }
+
+    private final CassandraAsyncExecutor queryExecutor;
+    private final PreparedStatement increaseStatement;
+    private final PreparedStatement decreaseStatement;
+    private final PreparedStatement getQuotaCurrentValueStatement;
+    private final PreparedStatement deleteQuotaCurrentValueStatement;
+
+    @Inject
+    public CassandraQuotaCurrentValueDao(CqlSession session) {
+        this.queryExecutor = new CassandraAsyncExecutor(session);
+        this.increaseStatement = session.prepare(increaseStatement().build());
+        this.decreaseStatement = session.prepare(decreaseStatement().build());
+        this.getQuotaCurrentValueStatement = session.prepare(getQuotaCurrentValueStatement().build());
+        this.deleteQuotaCurrentValueStatement = session.prepare(deleteQuotaCurrentValueStatement().build());
+    }
+
+    public Mono<Void> increase(QuotaKey quotaKey, long amount) {
```

Comment:
Counter write operations are peformed at Consistenc Level 1 and are non idempotent.

This makes it in practice prone to failure and not managing this properly can result (and did result with one of Linagora customers) in significant service disruption.

The best practice would be IMO to:
 - Log an error warning: `Failure when (in/de)creasing {component} {type} quota for {identifier}. Quota current value is thus errored and needs recomputation`
 - Ignore the error. We won't cash the service because of this.

Reply:
I think this could be handled in the class that call this method. We should not handle directly here.

Reply:
> We should not handle directly here.

Why?

Shall we expect the caller**s** to understand how  Cassandra counter works? 
Should the caller even knows that we are using Cassandra counters (ie should we create coupling?)?
Should we duplicate this bit of error handling?
Would moving the error handling bit make it more accurate?

Reply:
ok